### PR TITLE
add basic config struct to libkpod

### PR DESCRIPTION
**Pull request by 14rcole**
Add a basic config struct to libkpod so that it can be used in commands such as `kpod logs` and `kpod stats`.  This can be refined once kpod gets fleshed out more but it prevents having magic constants all over the place

### CODE REVIEW

**Comment by 14rcole** [comment]
@mheon @mrunalp PTAL

File: libkpod/config.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	Conmon                   string
+	ConmonEnv                []string
+	CgroupManager            string
+	DefaultTransport         string
+	InsecureRegistries       []string
+	SignaturePolicyPath      string
+}
+
+// DefaultConfig returns a Config struct with all options set to the default
+func DefaultConfig() *Config {
```
Comment (mheon) [inline libkpod/config.go]: I don't know if we want to bother with this at first. I'd be fine with leaving libkpod consumers to configure everything themselves and manage defaults independently. @mrunalp WDYT?

File: server/server.go [hunk 2]
```diff
@@ -510,7 +505,10 @@ func New(config *Config) (*Server, error) {
 		return nil, err
 	}
 
-	containerServer := libkpod.New(r, store, imageService, config.SignaturePolicyPath)
+	containerServer, err := libkpod.New(libkpod.DefaultConfig())
+	if err != nil {
```
Comment (TomSweeneyRedHat) [inline server/server.go]: Maybe I'm missing it, but don't you need to get the values of r, store, imageService and config.SignaturePolicyPath into the containerServer still?  Will the DefaultConfig() always fill the bill here?  And if not, do you still need to set them in this function?

File: server/server.go [hunk 3]
```diff
@@ -510,7 +505,10 @@ func New(config *Config) (*Server, error) {
 		return nil, err
 	}
 
-	containerServer := libkpod.New(r, store, imageService, config.SignaturePolicyPath)
+	containerServer, err := libkpod.New(libkpod.DefaultConfig())
```
Comment (mheon) [inline server/server.go]: Why default config here? Why not configure it from the config we're getting from file/command line?

File: libkpod/containerserver.go [hunk 4]
```diff
@@ -58,19 +58,38 @@ func (c *ContainerServer) ImageContext() *types.SystemContext {
...
+	imageService, err := storage.GetImageService(store, config.DefaultTransport, config.InsecureRegistries)
+	if err != nil {
+		return nil, err
+	}
+
+	runtime, err := oci.New(config.Runtime, config.RuntimeUntrustedWorkload, config.DefaultWorkloadTrust, config.Conmon, config.ConmonEnv, config.CgroupManager)
+	if err != nil {
+		return nil, err
+	}
 	return &ContainerServer{
```
Comment (mheon) [inline libkpod/containerserver.go]: We should store the Config inside the ContainerServer so we can look it up after creation time

File: server/server.go (hunk 3 above)
Comment (mheon) [inline server/server.go]: Also, since we're now setting up c/image and c/storage stuff inside libkpod, can probably remove that stuff from server.

File: server/server.go (hunk 2 above)
Comment (14rcole) [inline server/server.go]: The DefaultConfig() will provide some of those, others can be generated from the information within it

File: server/server.go (hunk 3 above)
Comment (14rcole) [inline server/server.go]: Good point

File: libkpod/containerserver.go (hunk 4 above)
Comment (14rcole) [inline libkpod/containerserver.go]: Works for me

File: libkpod/containerserver.go [hunk 5]
```diff
@@ -25,6 +25,7 @@ type ContainerServer struct {
 	imageContext       *types.SystemContext
 	stateLock          sync.Locker
 	state              *containerServerState
+	config             *Config
```
Comment (mheon) [inline libkpod/containerserver.go]: It doesn't look like you're assigning anything to this in New()

Comment (14rcole) [inline libkpod/containerserver.go]: Sorry about that, should be set now

File: libkpod/config.go (hunk 1 above)
Comment (rhatdan) [inline libkpod/config.go]: The default for a lot of thee fields should come out of the library. Containers/storage now has a /etc/containers/storage.conf which incidates the default place an admin will store configuration.  
Root directory below and RunRoot should definitely come from there.

I think we want to get similar defaults out of the containers/image at some point.  

We don't want users to have to configure buildah, skopeo, kpod, cri-o all differently.  We need to have defaults in the libraries so that they set it once and all of the commands follow the rules, unless the command itself wants to override.

Comment (14rcole) [inline libkpod/config.go]: Is there some way to read these files already built into containers/storage, containers/image, or crio?

Comment (mheon) [inline libkpod/config.go]: At this point, I think it would be better to just remove this, and trust kpod, buildah, and CRI-O to handle configuration and defaults themselves and assume they pass us a sane config. We can consider how best to integrate the c/storage and c/image default configs in a later PR.

**Comment by rhatdan** [comment]
I think store.DefaultStoreOptions has the data.  I think containers/image will just work if you specify no signature.  @runcom @mtrmac WDYT

**Comment by mtrmac** [comment]
> I think containers/image will just work if you specify no signature. @runcom @mtrmac WDYT

Is this referring to the `SignaturePolicyPath` field?

In short, if `SystemContext.SignaturePolicyPath` is empty, the library will read `/etc/containers/policy.json`.

In more detail: containers/image ’s `SystemContext` is designed designed so that _everything is optional_: a default-initialized value of that struct means “use appropriate defaults from the surronding environment”, including reading config files and environment variables. No caller is expected to know about every single member of `SystemContext`, and many simple callers should be able to pass `nil` instead of a pointer to their own context.

File: libkpod/config.go (hunk 1 above)
Comment (mrunalp) [inline libkpod/config.go]: Let us match the defaults everywhere and then allow overrides.

**Comment by 14rcole** [comment]
There seemed to be different opinions as to how the config should be handled, but here's what I've done:

- Pulled what data I could from store.DefaultStoreOptions to for DefaultConfig
-Added a global flag to kpod that accepts the path of a config file which can be used to modify configuration options

Let me know if I should change anything else

File: cmd/kpod/common.go [hunk 6]
```diff
@@ -31,3 +32,30 @@ func getStore(c *cli.Context) (storage.Store, error) {
...
+
+func getConfig(c *cli.Context) (*libkpod.Config, error) {
+	config := libkpod.DefaultConfig()
+	if c.GlobalIsSet("config") {
+		err := config.FromFile(c.String("config"))
+		if err != nil {
+			return config, err
+		}
+	}
+	if c.GlobalIsSet("root") {
```
Comment (mheon) [inline cmd/kpod/common.go]: If we're providing a recommended set of flags, there should be a flag array somewhere in here with them, so commands wanting to implement them can just append it into their flags.

Comment (mheon) [inline cmd/kpod/common.go]: It looks like they're global flags for the kpod command itself, but I still think they should be kept here - the code for parsing them and the flags themselves should live in the same place

File: libkpod/config.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"io/ioutil"
+
+	"github.com/BurntSushi/toml"
+	cstorage "github.com/containers/storage"
+)
+
+// Config stores the configuration options for libkpod
+type Config struct {
```
Comment (mheon) [inline libkpod/config.go]: Probably want to have TOML field name annotations to match CRI-O's config. Most of the options should be the same, so the same names/layout should be sufficient.

Eventually we can deduplicate and have CRI-O's config be a superset of this config.

File: libkpod/config.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	CgroupManager            string
+	DefaultTransport         string
+	InsecureRegistries       []string
+	SignaturePolicyPath      string
+}
+
+// FromFile populates the Config from the TOML-encoded file at the given path.
+// Returns errors encountered when reading or parsing the files, or nil
+// otherwise.
+func (c *Config) FromFile(path string) error {
```
Comment (14rcole) [inline libkpod/config.go]: I didn't think we couldn't simply because the CRI-O config struct contains embedded structs, which requires the use of another struct.  The config is marshalled into the toml struct and converted into the config struct, whereas in libkpod we can marshal directly (at least for now).

See https://github.com/kubernetes-incubator/cri-o/blob/master/server/config.go#L186

File: cmd/kpod/common.go (hunk 6 above)
Comment (14rcole) [inline cmd/kpod/common.go]: There's a set of global flags in main.go and every subcommand should have access to it, but why do we need to append the flags to each command if they can already be accessed globally?

File: libkpod/config.go (hunk 8 above)
Comment (mheon) [inline libkpod/config.go]: Maybe it would be simpler to just move most of the CRI-O config in here, then. We will eventually end up reusing every part except APIConfig, I imagine. We can move all the other structs in here, and have a separate version of the config that does not include APIConfig. @14rcole WDYT? If this ends up being harder than the way you're doing it now, I have no objections to this approach.

Comment (14rcole) [inline libkpod/config.go]: That's completely fine with me

File: cmd/kpod/common.go (hunk 6 above)
Comment (mheon) [inline cmd/kpod/common.go]: My preference would be to keep the config-related flags we're parsing here in this file, and then append them to the global flags list in kpod's main. That way, the flags we're parsing here are also present in this file, and all changes can happen in the same place.

Comment (14rcole) [inline cmd/kpod/common.go]: So everything that can be set in a config file can also be passed in as a flag?

Comment (mheon) [inline cmd/kpod/common.go]: Probably not everything, but most things - see mergeConfig in CRI-O. This doesn't have to happen yet, though.

**Comment by 14rcole** [comment]
The commit that I just added switches the various kpod commands to using getConfig() to get the store options.  The other commit in this PR previously failed only because getConfig() was unused

File: cmd/kpod/pull.go [hunk 9]
```diff
@@ -45,7 +45,11 @@ func pullCmd(c *cli.Context) error {
 	}
 	image := args[0]
 
-	store, err := getStore(c)
+	config, err := getConfig(c)
+	if err != nil {
+		return errors.Wrapf(err, "Could not get config")
```
Comment (mrunalp) [inline cmd/kpod/pull.go]: nit: Could -> could as you are wrapping and not logging.

File: cmd/kpod/common.go (hunk 6 above)
Comment (14rcole) [inline cmd/kpod/common.go]: I think I would rather make that a separate PR.  It would be a very easy one though

Comment (mheon) [inline cmd/kpod/common.go]: This one is large enough as it is, so it should definitely be done in a followon.

File: libkpod/config.go [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+
+	// PluginDir is where CNI plugin binaries are stored.
+	PluginDir string `toml:"plugin_dir"`
+}
+
+// tomlConfig is another way of looking at a Config, which is
+// TOML-friendly (it has all of the explicit tables). It's just used for
+// conversions.
+type tomlConfig struct {
+	Crio struct {
```
Comment (mheon) [inline libkpod/config.go]: I want to rename this to Kpod, but that would break compatability with existing CRI-O configs, so let's leave as is for now.

Comment (14rcole) [inline libkpod/config.go]: I could go through and change the occurences.  It's just a bit of search and replace

Comment (mheon) [inline libkpod/config.go]: But the existing CRI-O configs generated prior use `Crio` - I don't know how large our install base is, but we should try to avoid breaking things anyways.

**Comment by mheon** [comment]
LGTM if tests pass

File: libkpod/config.go [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+	t.Crio.RootConfig = c.RootConfig
+	t.Crio.Runtime.RuntimeConfig = c.RuntimeConfig
+	t.Crio.Image.ImageConfig = c.ImageConfig
+	t.Crio.Network.NetworkConfig = c.NetworkConfig
+}
+
+// FromFile populates the Config from the TOML-encoded file at the given path.
+// Returns errors encountered when reading or parsing the files, or nil
+// otherwise.
+func (c *Config) FromFile(path string) error {
```
Comment (mtrmac) [inline libkpod/config.go]: Would it make sense to call this `UpdateFromFile` instead? `FromFile` would commonly mean _only_ from that file, but this seems to preserve a lot (potentially everything?) from the original contents of `c`.

(And if this intended to preserve nothing from `c`, it should be a stand-alone function / constructor, not a method on `c`.)

Comment (14rcole) [inline libkpod/config.go]: That would make sense.  It should preserve data from c so that any fields not set in the file can remain as-is so I'll change that

**Comment by 14rcole** [comment]
@mrunalp Good to merge?

**Comment by 14rcole** [comment]
Should getConfig() automatically attempt to read /etc/containers/storage.conf if no config file is provided?

**Comment by rhatdan** [comment]
That should happen automatically. I have just opened a pull request to fix crio use of storage.conf.
https://github.com/kubernetes-incubator/cri-o/pull/701

**Comment by mrunalp** [comment]
LGTM
